Enable credential provider authentication for self-hosted mysql DB

ssrangisetti opened the pull request.

saisandeep-flipkart committed:
Enable credential provider authentication for self-hosted mysql DB

saisandeep-flipkart committed:
refactored New function to NewClient

saisandeep-flipkart committed:
using opentelemetry http wrapper and using http client's in-built timeout

amanhigh commented on lib/config/fileconf.go:
```diff
@@ -1540,6 +1540,20 @@ type Databases struct {
+	TokenConfig TokenConfig `yaml:"token,omitempty"`
+}
+
+type TokenConfig struct {
```
Can we have defaults so client can get started with minimal config. Eg. https://github.com/caarlos0/env or VIPER

amanhigh commented on lib/config/fileconf.go:
```diff
@@ -1540,6 +1540,20 @@ type Databases struct {
+
+type TokenConfig struct {
+	Enabled           types.BoolOption `yaml:"enabled"`
+	UrlTmpl           string           `yaml:"url_tmpl"`
```
Can use full names which are self explanatory. Also TokenSourceConfig or something else ?

amanhigh commented on lib/token/auth.go:
```diff
@@ -0,0 +1,32 @@
+	AuthCreators[NoneAuthScheme] = newNoAuthInjector
+}
+
+type AuthInjector interface {
```
Can we vet interface against popular cloud Identity services eg. GCP, Azure etc

amanhigh commented on lib/token/client.go:
```diff
@@ -0,0 +1,125 @@
+	Username string
+}
+
+func NewClient(config ClientConfig) *Client {
```
Can use a popular client like https://github.com/go-resty/resty

amanhigh commented on lib/token/client_test.go:
```diff
@@ -0,0 +1,163 @@
+	"time"
+)
+
+func TestNewClient(t *testing.T) {
```
Can avoid test if we use existing client.
https://github.com/go-resty/resty

amanhigh commented on lib/config/configuration_test.go:
```diff
@@ -3403,3 +3404,64 @@ func TestApplyDiscoveryConfig(t *testing.T) {
 	}
 }
+
+func TestApplyFileConfig_DbTokenAuth_errors(t *testing.T) {
```
Can add tests for missing mandatory fields.

amanhigh commented on lib/config/configuration_test.go:
```diff
@@ -3403,3 +3404,64 @@ func TestApplyDiscoveryConfig(t *testing.T) {
+func TestApplyFileConfig_DbTokenAuth(t *testing.T) {
+	defaultCfg := service.MakeDefaultConfig()
+	err := ApplyFileConfig(&FileConfig{
+		Databases: Databases{
```
Can add happy test for Full Config (With Optionals like timeout etc) , Minimal Config (Bare minimum required).

amanhigh reviewed (commented):
Few other observations which may help user by time feature is compleete
- Diagram explaning interaction with Tokensource in Readme
- Easy setup in local and trial with dummy token source.

saisandeep-flipkart committed:
Addressed some of the review comments

Renamed token, tokenConfig, tokenClient to tokensource, tokenSourceConfig, tokenSourceClient
Using go-resty instead of default http client
AuthInjector can inject more types of authentications like mTLS
Added default values and corresponding tests for token source config

amanhigh commented on lib/tokensource/client.go:
```diff
@@ -0,0 +1,102 @@
+
+func (c *Client) GetCredentials(ctx context.Context, dbName, username, token string) (string, string, error) {
+	buf := &bytes.Buffer{}
+	if err := c.urlTemplate.Execute(buf, urlTmplParams{DBName: dbName, Token: token, Username: username}); err != nil {
```
Optional: move Url Subtitution to buildUrl Helper Function

amanhigh commented on lib/tokensource/client.go:
```diff
@@ -0,0 +1,102 @@
+	}
+}
+
+func (c *ClientConfig) CheckAndSetDefaults() {
```
Should this be called in Constructor NewClient so no one missises calling it ?

amanhigh commented on lib/config/configuration.go:
```diff
@@ -1296,21 +1297,24 @@ func applyKubeConfig(fc *FileConfig, cfg *service.Config) error {
 		var err error
-		cfg.Databases.TokenConfig.UrlTmpl, err = template.New("url").Parse(tokenConfig.UrlTmpl)
+		cfg.Databases.TokenSourceConfig.Enabled = tokenConfig.Enabled.Value
+		if tokenConfig.UrlTemplate == "" {
```
Can some of these checks be moved to CheckAndSetDefault for encapsulation ?

amanhigh reviewed (approved):
Overall Looks Good Few Comments.

saisandeep-flipkart committed:
Addressed some of the review comments

Calling CheckAndSetDefaults in NewClient
Moved url substitution to a new function

amanhigh reviewed (approved):
Looks Good
